ntoskrnl/io/iomgr/irq.c

```diff
@@ -153,25 +157,22 @@ IoDisconnectInterrupt(PKINTERRUPT InterruptObject)
     KeDisconnectInterrupt(&IoInterrupt->FirstInterrupt);
 
     /* Now disconnect the others */
-    for (i = 0; i < KeNumberProcessors; i++)
+    for (i = 0; i < _countof(IoInterrupt->Interrupt) && IoInterrupt->Interrupt[i]; ++i)
     {
-        /* Make sure one was registered */
-        if (IoInterrupt->Interrupt[i])
-        {
-            /* Disconnect it */
-            KeDisconnectInterrupt(&InterruptObject[i]);
-        }
+        /* Disconnect it */
+        KeDisconnectInterrupt(IoInterrupt->Interrupt[i]);
     }
 
     /* Free the I/O Interrupt */
-    ExFreePool(IoInterrupt); // ExFreePoolWithTag(IoInterrupt, TAG_KINTERRUPT);
+    ExFreePoolWithTag(IoInterrupt, TAG_KINTERRUPT);
```

Comment:
Is that freeing the pair of that allocation in io.h ?
![image](https://github.com/reactos/reactos/assets/33393466/abe9701b-5b40-460f-9ea9-5ddadbafeded)
Just an assumption, because the same tag is used over there.

Reply:
what do you mean "the pair" ?

Reply:
> Is that freeing the pair of that allocation in io.h ?

Yes, obviously.

> why the correct freeing with the tag was inside the code already, but commented.

Ask the author, not me.

> Do your other changes contribute to the possibility of doing the right thing now?

Maybe the `KeDisconnectInterrupt()` and/or `SpinLock` fixes help (to not corrupt memory?)?

Reply:
I mean with "the pair" that it **at first glance** seems there is no other alloc belonging to that free.
![image](https://github.com/reactos/reactos/assets/33393466/992398f0-f63f-492f-b05e-0ee4e3a6ca6c)
That means at least as long as the define would be structurally used.
But grepping for RSIK seems to only give that unique hit also:
![image](https://github.com/reactos/reactos/assets/33393466/c164593a-1770-480a-9a51-7640dba0544f)
Really strange that it was allocated with the tag, but freed without.

Reply:
Ah yes, this was me in commit 24a14abf2 (r57412) who introduced a bunch of s/ExFreePool/ExFreePoolWithTag changes, and this one was commented out. While I don't remember the details why I commented only that one, it's quite probably because i was basically hitting a pool corruption here, due to the bug @SergeGautherie fixed in the disconnect loop.